src/platforms/esp32/components/avm_sys/platform_nifs.c

```diff
@@ -435,27 +449,156 @@ static term nif_esp_sleep_enable_ext1_wakeup(Context *ctx, int argc, term argv[]
+                                                                                                               \
+        mbedtls_##ALGORITHM##_init(&md_ctx);                                                                   \
+        mbedtls_##ALGORITHM##_starts##SUFFIX(&md_ctx, IS_OTHER);                                               \
+                                                                                                               \
+        InteropFunctionResult result = interop_iolist_fold(data, ALGORITHM##_hash_fold_fun_##IS_OTHER, (void *) &md_ctx); \
+        if (UNLIKELY(result != InteropOk)) {                                                                   \
+            return false;                                                                                      \
+        }                                                                                                      \
+                                                                                                               \
+        if (UNLIKELY(mbedtls_##ALGORITHM##_finish##SUFFIX(&md_ctx, dst) != 0)) {                               \
+            return false;                                                                                      \
+        }                                                                                                      \
+                                                                                                               \
+        return true;                                                                                           \
+    }
+
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 0, 0)
+
+DEFINE_DO_HASH(md5, _ret)
+DEFINE_DO_HASH(sha1, _ret)
+DEFINE_DO_HASH2(sha256, _ret, true)
+DEFINE_DO_HASH2(sha256, _ret, false)
+DEFINE_DO_HASH2(sha512, _ret, true)
+DEFINE_DO_HASH2(sha512, _ret, false)
+
+#else
+
+DEFINE_DO_HASH(md5, )
+DEFINE_DO_HASH(sha1, )
+DEFINE_DO_HASH2(sha256, , true)
+DEFINE_DO_HASH2(sha256, , false)
+DEFINE_DO_HASH2(sha512, , true)
+DEFINE_DO_HASH2(sha512, , false)
+
+#endif
+
+static term nif_crypto_hash(Context *ctx, int argc, term argv[])
 {
     UNUSED(argc);
-    term data = argv[0];
-    VALIDATE_VALUE(data, term_is_binary);
-
-    unsigned char digest[MD5_DIGEST_LENGTH];
-    struct MD5Context md5;
-    #if __has_include (<esp_idf_version.h>) && ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 3, 0) && CONFIG_IDF_TARGET_ESP32
-    esp_rom_md5_init(&md5);
-    esp_rom_md5_update(&md5, (const unsigned char *) term_binary_data(data), term_binary_size(data));
-    esp_rom_md5_final(digest, &md5);
-    #else
-    MD5Init(&md5);
-    MD5Update(&md5, (const unsigned char *) term_binary_data(data), term_binary_size(data));
-    MD5Final(digest, &md5);
-    #endif
-    if (UNLIKELY(memory_ensure_free(ctx, term_binary_heap_size(MD5_DIGEST_LENGTH)) != MEMORY_GC_OK)) {
+    term type = argv[0];
+    VALIDATE_VALUE(type, term_is_atom);
```

Comment:
extreme nitpicking: `interop_atom_term_select_int` already does this for us, however this behavior is not explicitly mentioned in our documentation.